### ISSUE TITLE
no matching function for call to ‘max(double, ceres::TinySolver ... ::Scalar)’

### ISSUE THREAD

**Issue by lacava** [heal-research/operon#6]
I get this error when I run cmake with these options:

```
cmake .. -DCMAKE_BUILD_TYPE=Release  -DBUILD_PYBIND=ON -DUSE_OPENLIBM=ON -DUSE_SINGLE_PRECISION=ON -DCERES_TINY_SOLVER=ON
```

as recommended in issue #5.

```
[ 70%] Building CXX object CMakeFiles/pyoperon.dir/src/python/eval.cpp.o
/usr/bin/gcc-9 -DCERES_TINY_SOLVER -DUSE_SINGLE_PRECISION -Dpyoperon_EXPORTS -I/home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/python -isystem /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/include/operon -isystem /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/thirdparty/GSL/include -isystem /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/thirdparty/CSV/include -isystem /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/thirdparty/nanobench/include -isystem /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/thirdparty/xxhash -isystem /home/bill/anaconda3/envs/regression-benchmarks/include/python3.9 -isystem /home/bill/anaconda3/envs/regression-benchmarks/include/eigen3 -isystem /home/bill/anaconda3/envs/regression-benchmarks/include -O3 -DNDEBUG -fPIC -fvisibility=hidden -Wall -Werror -Wextra -Wshadow -Wsign-compare -Wwrite-strings -Wpointer-arith -Winit-self -Wconversion -Wno-sign-conversion -O3 -g -fPIC -march=native -flto -fno-fat-lto-objects -std=gnu++17 -o CMakeFiles/pyoperon.dir/src/python/eval.cpp.o -c /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/src/python/eval.cpp
In file included from /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/include/operon/nnls/nnls.hpp:25,
                 from /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/include/operon/operators/evaluator.hpp:26,
                 from /home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/src/python/eval.cpp:8:
/home/bill/anaconda3/envs/regression-benchmarks/include/ceres/tiny_solver.h: In instantiation of ‘const ceres::TinySolver<Function, LinearSolver>::Summary& ceres::TinySolver<Function, LinearSolver>::Solve(const Function&, ceres::TinySolver<Function, LinearSolver>::Parameters*) [with Function = Operon::DynamicAutoDiffCostFunction<Operon::ResidualEvaluator, ceres::Jet<float, 4>, 0>; LinearSolver = Eigen::LDLT<Eigen::Matrix<float, -1, -1>, 1>; ceres::TinySolver<Function, LinearSolver>::Parameters = Eigen::Matrix<float, -1, 1>; typename Function::Scalar = float]’:
/home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/include/operon/nnls/nnls.hpp:115:17:   required from ‘ceres::Solver::Summary Operon::OptimizeTiny(Operon::Tree&, const Operon::Dataset&, gsl::span<const float>, Operon::Range, size_t, bool, bool) [with bool autodiff = true; size_t = long unsigned int]’
/home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/include/operon/nnls/nnls.hpp:132:30:   required from ‘auto Operon::OptimizeTinyAutodiff(Args&& ...) [with Args = {Operon::Tree&, const Operon::Dataset&, const gsl::span<const float, 18446744073709551615>&, const Operon::Range&, long unsigned int&, bool&, bool&}]’
/home/bill/projects/symbolic-regression/analysis/experiment/methods/src/operon/include/operon/operators/evaluator.hpp:35:110:   required from here
/home/bill/anaconda3/envs/regression-benchmarks/include/ceres/tiny_solver.h:285:25: error: no matching function for call to ‘max(double, ceres::TinySolver<Operon::DynamicAutoDiffCostFunction<Operon::ResidualEvaluator, ceres::Jet<float, 4>, 0> >::Scalar)’
  285 |         u = u * std::max(1 / 3., 1 - tmp * tmp * tmp);
```

It goes away if I set `-DCERES_TINY_SOLVER=OFF`.

**Comment by foolnotion** [comment]
Unfortunately this is a bug in the tiny solver (see [here](https://github.com/ceres-solver/ceres-solver/commit/e84cf10e13633618a780543e83c117a84316b790)), the plan is to separate tiny solver from ceres and use the latest version as thirdparty. Expect it to be fixed in the next few days in Operon.

**Comment by foolnotion** [comment]
This should be now fixed with 9da820fc4399b3142fdf8a601cf4de34b381d629. The ceres tiny solver is now included standalone and will be used if `CERES_TINY_SOLVER=ON` or if `Ceres` is not found.

**Comment by lacava** [comment]
thanks, that seemed to fix the ceres issue.
